### PR TITLE
quant_Bigramoptimizer-spikinghashnet submission

### PR DESCRIPTION
**Pull request by Laalinibh**
Model variant: BigramHashEmbedding (hash_size=307) with spiking causal self-attention and Muon+Adam optimizer split.
Logs are one run per file:
[train_v0.txt](vscode-file://vscode-app/Applications/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-browser/workbench/workbench.html)
[train_v1.txt](vscode-file://vscode-app/Applications/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-browser/workbench/workbench.html)
[train_v2.txt](vscode-file://vscode-app/Applications/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-browser/workbench/workbench.html)
Final exact metrics (from terminal logs):
train_v0: val_loss=3.03380227, val_bpb=1.79678859
train_v1: val_loss=3.87858984, val_bpb=2.29711937
train_v2: val_loss=2.89974137, val_bpb=1.71739017
3-run mean: val_loss=3.27071116, val_bpb=1.93709938 (reported in

### CODE REVIEW

**Review by Laalinibh — COMMENTED** [review]
his submission fixes a training bug in `train_gpt.py` for models using `BigramHashEmbedding`, and documents the full architecture used for the submission.

## Architecture

### BigramHashEmbedding
- Uses two independent hash tables:
  - `unigram_table` with shape `(hash_size, dim)`
  - `bigram_table` with shape `(hash_size, dim)`
- For each token, embedding is computed as:
  - `unigram_table[token_id % hash_size]`
  - `bigram_table[(prev_id * 31 + token_id) % hash_size]`
- This reduces embedding parameters from `vocab_size * dim` to `2 * hash_size * dim`.
- When `tie_embeddings=True`, the output projection still uses `tok_emb.weight`, which maps to `unigram_table.weight`.

### SpikingCausalSelfAttention
- Self-attention is implemented as a spiking causal attention block.
- Queries, keys, values, and gates are generated by linear projections.
- The query stream is gated and scaled by a learned `q_gain` parameter.
- The attention operation uses `F.scaled_dot_product_attention(..., is_causal=True)`.
- The module tracks a membrane-style state and detaches it between forward passes.

### Block design
- Each transformer block contains:
  - `RMSNorm` before attention and MLP
  - `SpikingCausalSelfAttention`
  - `MLP` with `LeakyReLU²` activation
  - Learned residual mixing via `resid_mix`
  - Separate scaling parameters for attention and MLP outputs
- The architecture is a modified transformer with spiking attention and hash-based token embeddings.

### Training and optimizer split
- Token embeddings are optimized separately from matrix and scalar parameters.
- `optimizer_tok` handles token embedding weights.
- `optimizer_muon` handles matrix parameters in transformer blocks.
- `optimizer_scalar` handles scalar and control parameters.
